Add concept exercise objects (#1160)
exercism/javascript: concepts/objects/about.md

# About

## Explanation

Besides primitive data types like `number` and `string`, there is another important data type in JavaScript called `object`.
Objects are collections of key-value pairs.

> In other languages they are referred to as maps or dictionaries and often times values need to have the same data type.

Reviewer: ```suggestion
In other languages they are referred to as maps or dictionaries and often times all values need to have the same data type.
```

Reviewer: My initial guess would have been that they're more similar to structs or data classes in other languages, which can have varying value types. Perhaps you could briefly mention how those two relate?

Reviewer 2: Also, on this line "often times" isn't familiar sounding English to me. Just "often" without the "times" is clearer 🙂

Reviewer: I think it's a US phrase mostly

Author: @SaschaMann you are totally right with that first comment, in my head it was clear that we will cover that struct/class aspect of objects elsewhere but I forgot to also make this clear for the reader. Rewrote that section. Removed the "times" as well.

In JavaScript, only the type of the key is restricted, it has to be a string.
The values inside one object can have different types.

> They can be primitive types like numbers but also arrays, other objects or even functions.

Author: I meant is as "but also entities like arrays" not as "but also types like arrays". I think it is ok as is. The "object basics" concept is not the right point for the "well actually functions are objects" point.

## Creating an Object

> You create an object (literal) with curly brackets.

Reviewer: Does the student know what `(literal)` refers to at this point in the track?

If not, perhaps adding a brief footnote or explanation would prevent some web searching.

Author: Good point. Jeremy recommended to use links in these cases where we use terms that someone that already knows programming should know but in reality might not know.

You can also directly include some entries (key-value pairs).
For that, state the key first, followed by a colon and the value.

```javascript
const emptyObject = {};

const obj = {
  favoriteNumber: 42,
  greeting: 'Hello World',
  useGreeting: true,
  address: {
    street: 'Trincomalee Highway',
    city: 'Batticaloa',
  },
  fruits: ['melon', 'papaya'],
  addNumbers: function (a, b) {
    return a + b;
  },
};
```

> The trailing comma after the last entry is optional in JavaScript.

Reviewer: It may be a good idea to mention why one would want to add a trailing comma (better diffs etc).

Author: Provided a link where this is spelled out. I think that is sufficient.

You might wonder why the keys are not wrapped in quotation marks although they are supposed to be strings.
This is a short-hand notation.
If the key follows the naming rules for a JavaScript [identifier][mdn-identifier], you can omit the quotation marks.
For keys with special characters in the name, you need to apply the usual string notation.

```javascript
const obj = {
  '1keyStartsWithNumber': '...',
  'key/with/slashes': '...',
  'key-with-dashes': '...',
  'key with spaces': '...',
  '#&()[]{}èä樹keyWithSpecialChars': '...',
};
```

Often times you want to combine existing variables into an object.
This would often lead to key-value pairs like `name: name`.

> JavaScript has a short-hand notation that allows to just write `name` instead.

Reviewer: ```suggestion
JavaScript has a short-hand notation that allows just writing `name` instead.
```

`allow` is followed by a gerund, I think.

There is also a short-hand notation for defining a function in an object.

> There you can omit the colon and the `function` keyword.

Reviewer 3: Might not be the place to mention this, but the short-hand notation is **not** equivalent to the key-value notation. The difference is small, but there:

```javascript
const appleTree = {
  kind: 'apple',
  getParent: function() {
    return super.toString();
  }
};
// Uncaught SyntaxError: 'super' keyword unexpected here

const pearTree = {
  kind: 'pear',
  getParent() {
    return super.toString();
  }
};
// this works

// You can then declare things on its prototype
Object.setPrototypeOf(pearTree, { toString() { return "BigTree" } })
pearTree.getParent()
// => "BigTree"
```

It's a weird edgy case, but it matters as soon as you're going to type to _copy_ that method you've declared or if you're trying to write mixins.

Reviewer 3: There's another _difference_ between these:

```javascript
const obj = {
  calcSum: function calc(a, b) {
    return a + b;
  },
};

obj.calcSum.name
// => "calc"

const obj2 = {
  calcSum: function (a, b) {
    return a + b;
  },
};

obj2.calcSum.name
// => "calcSum"

const obj3 = {
  calcSum(a, b) {
    return a + b;
  },
};

obj3.calcSum.name
// => "calcSum"
```

It's pretty insignificant, but 🤷🏽

Author: Re the first issue: 
Interesting, I didn't know and it was not mentioned [on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions#description). Do you know why there is a difference or can you provide some link where I could read up on this? Tbh I would not hold off merging the PR until I got in. But I can put it into the design document as improvement idea. It might be a good fit once we have the special "sections" for about.md where you can mark something as advanced content.

Re the second issue:
True, but this is more about how function names are determined than about the shorthand notation imo. It could go into the functions concept but even there we have a lot more basic ground to cover first.

Its not worth holding off this PR for any of those. I would put both things into the design.md improvement ideas section of the respective exercise. They might be a good fit once we have the special "sections" for about.md where you can mark something as advanced content.

```javascript
const x = 1;
const y = 2;

const obj = {
  x: x,
  y: y,
  calcSum: function (a, b) {
    return a + b;
  },
};

// Can be shortened to ...
const obj = {
  x,
  y,
  calcSum(a, b) {
    return a + b;
  },
};
```

If you want to create an object literal with a key name that is stored in a variable or derived via an expression, you need to wrap the variable or expression in square brackets.

```javascript
const key = 'greeting';

const obj = {
  [key]: 'hello world',
};

> ```

Reviewer 3: Might be worth showing the result

## Retrieving a Value

> There are two ways to retrieve the value for a given key, dot notation and bracket notation.

Reviewer 3: We should definitely NOT mention this, but there are more than two ways.

```javascript
const obj = { greeting: 'hello world' }

const descriptor = Object.getOwnPropertyDescriptor(obj , 'greeting')
// => {value: "hello world", writable: true, enumerable: true, configurable: true}

descriptor.value
// => "hello world"
```

And naturally by way of destructuring.

There's even `Object.values`:

```javascript
const obj = { foo: 'bar', greeting: 'hello world' }

const values = Object.values(obj)
const index = Object.getOwnPropertyNames(obj).indexOf('greeting')

values[index]
// => "hello world"
```

By placing these in this comment on this PR, someone, someday might find them and be sad or happy.

Author: "There are some more ways ... " probably applies to every other thing I cover in all the JavaScript concepts. :wink:

```javascript
const obj = { greeting: 'hello world' };

obj.greeting;
// => hello world

obj['greeting'];
// => hello world

// Bracket notation also works with variables.
const key = 'greeting';
obj[key];
// => hello world
```

Using the dot notation as a short-hand has the same restriction as omitting the quotation marks.
It only works if the key follows the identifier naming rules.

If you try to retrieve a key that does not exist in the object, JavaScript returns `undefined`.

> See the [Null and Undefined Concept][concept-null-undefined] for more details on this.

Reviewer: ```suggestion
See [Null and Undefined][concept-null-undefined] for more details on this.
```

You can chain the keys if you want to retrieve a value from a nested object.

```javascript
const obj = {
  address: {
    street: 'Trincomalee Highway',
    city: 'Batticaloa',
  },
};

obj.address.city;
// => 'Batticaloa'

obj['address']['city'];
// => 'Batticaloa'

> ```

Reviewer 3: Perhaps worth mentioning that you _could_ mix and match?

```javascript
obj.address['city']
```

## Adding or Changing a Value

You can add or change a value using the assignment operator `=`.
Again, there is dot and bracket notation available.

```javascript
const obj = { greeting: 'hello world' };

obj.greeting = 'Hi there!';
obj['greeting'] = 'Welcome.';

obj.newKey1 = 'new value 1';
obj['new key 2'] = 'new value 2';

const key = 'new key 3';
obj[key] = 'new value 3';

> ```

Reviewer 3: If you add the property using `Object.create` or like this:

```javascript
const obj = {}
Object.defineProperty(obj, 'greeting', { value: 'hello world', configurable: true })
// => {greeting: "hello world"}
```

You can actually then do:

```javascript
Object.defineProperty(obj, 'greeting', { value: 'goodbye world' })
// => {greeting: "goodbye world"}
```

`defineProperty` etc is probably something we want somewhere but not here, but I don't know where so that's why I put it here so I hope it gives you inspiration.

Author: Totally agree it should go in somewhere, that is why I wrote "Helper methods like `Object.assign` etc." should be covered in the about.md in the future in the design document. I rather wanted to move on to other concepts at some point. So in a next iteration on this I would see these helpers being explained in the `objects` about.md and then further down the road they could even be moved into some "objects advanced" or whatever exercise if we find more related things we maybe want to cover with real exercises and not just mention in about.md.

## Deleting an Entry

You can delete a key-value pair from an object using the `delete` keyword.

```javascript
const obj = {
  key1: 'value1',
  key2: 'value2',
};

delete obj.key1;
delete obj['key2'];
```

> Note that although `undefined` is returned for non-existing keys, setting a key to `undefined` does not remove the entry.

Reviewer 3: Very good advice!

Might be worth showing `hasOwnProperty` to validate something is gone, given that you show it below?

Author: I think just showing it below is enough here.

Always use `delete` instead.

## Checking Whether a Key Exists

You can check whether a certain key exists in an object with the `hasOwnProperty` method.

```javascript
const obj = { greeting: 'hello world' };

obj.hasOwnProperty('greeting');
// => true

obj.hasOwnProperty('age');
// => false
```

`hasOwnProperty` returns `false` for [inherited keys][concept-inheritance]. This is usually the desired behavior.

> If you want include inherited keys in the existence check you can use the [`in` operator][mdn-in-operator] instead.

Reviewer 3: You might want to mention `getOwnPropertyNames` which are all the properties `hasOwnProperty` is true for.

Author: Same category as "defineProperty", I added it the improvement ideas list.

## Looping Through an Object

> There is a special `for...in` loop to iterate over all keys of an object.

Reviewer: Are objects in JS iterables? Would `for…of` work too?

Reviewer 3: `for...in` and `for...of` are not equivalent, and the concept of enumerable is not the same as iterable, in JavaScript (@junedev all this might be worth mentioning, somewhere).

JavaScript objects have _enumerable properties_. They are enumerable by default, and can be made non enumerable using `Object.create` or `Object.defineProperty` setting `{ enumerable: false }`.

They are _only_ iterable if there is `Symbol.iterator` defined (as a method):

```javascript
const iterable1 = {};

iterable1[Symbol.iterator] = function* () {
  yield 1;
  yield 2;
  yield 3;
};

console.log([...iterable1]);
// [1, 2, 3]
```

[Array's prototype has it defined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/@@iterator), but Object's prototype does not. _Iterability_ is **only** used by `for ... of`. The async variant is `Symbol.asyncIterator` and is used by **for-await-of**.

Author: @SleeplessByte I would not go into the details of this in this concept. Left a note (and the link to this comment) on our concept tracking issue. https://github.com/exercism/javascript/issues/951#issuecomment-864452573

Reviewer: Depending on what comes first in the track, this may be worth mentioning, i.e. if `for..of` is explained in a previous exercise. That way, the student knows to watch out that `for..of` won't work here.

Thanks for the explanation :)

Author: `for...of` comes later in our current unlock flow but we treat about.md as independent from that flow. So your point is totally valid. I added a small part on this. I used a link to the MDN docs for now. Once we have our own concept on enumerable and iterable we can link that instead.

```javascript
const obj = {
  name: 'Ali',
  age: 65,
};

for (let key in obj) {
  console.log(key, obj[key]);
}
// name Ali
// age 65
```

It might seem like `for...in` always visits the keys in the order in which they appear or were added to the object (insertion order).
Nevertheless you should not rely on this because there are special cases where the order will differ from the insertion order.
Always write your code as if the order would be arbitrary.
If you need a reliable order, use a [`Map`][jsinfo-map] or a nested [array][concept-arrays] instead of an object.

> The for-in loop has a pitfall in case the object has [inherited keys][concept-inheritance].

Reviewer: ```suggestion
The `for…in` loop has a pitfall in case the object has [inherited keys][concept-inheritance].
```

Reviewer: Not sure which one is the common name, but I think consistency might be nicer. Same applies to the introduction.

> Those will also be visited by the for-in loop.

Reviewer: ```suggestion
Those will also be visited by the `for…in` loop.
```

> You can avoid this by skipping over them like shown below.

Reviewer 3: Perhaps it should be mentioned that you almost _always_ want this.

Author: Good point, I changed the wording accordingly.

# About

## Explanation

Besides primitive data types like `number` and `string`, there is another important data type in JavaScript called `object`.
Objects are collections of key-value pairs.

> As such, they can be used as what is referred to as maps or dictionaries in other languages.

Reviewer: ```suggestion
As such, they can be used as what is often referred to as maps or dictionaries in other languages.
```

structs in other languages are also sometimes used purely as a collection. It's rare but imo making this statement slightly weaker clears up that it doesn't _always_ translate to a map or dictionary even if in many languages it does.

The difference to classes with methods in the context of this exercise is clearer to me now, though, so I think it'd be fine as it is too.

Author: fixed in all three places

In other languages, all values in a map often need to have the same data type.
In JavaScript, only the type of the key is restricted: it has to be a string.
The values inside one object can have different types.
They can be primitive types like numbers but also arrays, other objects or even functions.
This makes objects very versatile so that they are also key entities for [object-oriented programming][oop] (OOP) in JavaScript.

> In the following we will focus on objects as collections or maps. Other use cases of objects will be covered in other concepts, see e.g., [Classes][concept-classes].

Reviewer: ```suggestion
In the following we will focus on objects as collections or maps. Other use cases of objects will be covered in other concepts, see e.g. [Classes][concept-classes].
```

Author: @SaschaMann Are you sure about this? I read [here](https://exercism.lol/docs/building/markdown/style-guide) that we use American English and from what read there you always put in the comma. 🤔

Reviewer: I have never seen a comma after `e.g.`, hence my suggestion, but apparently you're right and American English puts a comma there indeed. The doc you linked to gets that wrong too funnily enough.

The more you know... :D
